spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -366,15 +370,55 @@ public void pruneColumns(StructType requestedSchema) {
 
   private Schema schemaWithMetadataColumns() {
     // metadata columns
-    List<Types.NestedField> fields =
+    List<Types.NestedField> metadataFields =
         metaColumns.stream()
             .distinct()
             .map(name -> MetadataColumns.metadataColumn(table, name))
             .collect(Collectors.toList());
-    Schema meta = new Schema(fields);
+    // only calculate potential column id collision if metadata column was requested
+    Schema metadataSchema =
+        metadataFields.isEmpty()
+            ? new Schema(metadataFields)
+            : deduplicateSchemaIds(metadataFields);
 
     // schema or rows returned by readers
-    return TypeUtil.join(schema, meta);
+    return TypeUtil.join(schema, metadataSchema);
+  }
+
+  private Schema deduplicateSchemaIds(List<Types.NestedField> metaColumnFields) {
+    Map<Integer, Types.NestedField> indexedMetadataColumnFields =
+        TypeUtil.indexById(Types.StructType.of(metaColumnFields));
+
+    // Calculate field ids to reassign from nested partition field
+    Set<Integer> idsToReassign =
```

Comment:
Hm, I wonder could we just go through the metadataColumnFields until we find  id = PARTITION_COLUMN_ID, and then index that struct only?

Reply:
I think we can, but do you think it would be a bit less future proof? Right we are reassigning anything outside metadata column which have reserved column id, which include the partition struct.